File: player/raspberrypi/cms_control_client.py
"""
CMS control hub client for Raspberry Pi player.

Connects to `/hubs/cms` on the CCMS backend using the screen's API key as
the `access_token` query parameter. Receives `command` and `playlist_updated`
events from the dashboard; sends acknowledgements back via `AckCommand`.

Paired with:
  - remote_command_handler.RemoteCommandHandler — translates commands into mpv ops
  - cms_pair.py — CLI to claim a 6-char pairing code and persist credentials

This module is additive: it does NOT replace the existing marketplace
SignalR connection (`/hubs/playback`). Both run concurrently; marketplace
screens simply never receive commands because no CmsOwner owns them.
"""

import logging
from typing import Callable, Optional, Dict, Any

try:
    from signalrcore.hub_connection_builder import HubConnectionBuilder
except ImportError:
    HubConnectionBuilder = None

logger = logging.getLogger(__name__)


class CmsControlClient:
    """Thin SignalR wrapper for the CMS control hub."""

    def __init__(self, api_url: str, screen_id: str, api_key: str):
        self.api_url = api_url.rstrip("/")
        self.screen_id = screen_id
        self.api_key = api_key
        self._conn = None
        self._connected = False
        self._on_command: Optional[Callable[[Dict[str, Any]], None]] = None
        self._on_playlist_updated: Optional[Callable[[Dict[str, Any]], None]] = None

    def set_on_command(self, cb: Callable[[Dict[str, Any]], None]) -> None:
        self._on_command = cb

    def set_on_playlist_updated(self, cb: Callable[[Dict[str, Any]], None]) -> None:
        self._on_playlist_updated = cb

    def connect(self) -> bool:
        if HubConnectionBuilder is None:
            logger.error("signalrcore not installed — CMS hub disabled")
            return False
        try:
            hub_url = f"{self.api_url}/hubs/cms?access_token={self.api_key}"
            logger.info("Connecting to CMS hub: %s/hubs/cms", self.api_url)
            self._conn = (
                HubConnectionBuilder()
                .with_url(hub_url)
                .configure_logging(logging.WARNING)
                .with_automatic_reconnect({
                    "type": "raw",
                    "keep_alive_interval": 10,
                    "reconnect_interval": 5,
                    "max_attempts": 20,
                })
                .build()
            )
            self._conn.on_open(self._on_open)
            self._conn.on_close(self._on_close)
            self._conn.on_error(lambda e: logger.error("CMS hub error: %s", e))
            self._conn.on("command", self._handle_command)
            self._conn.on("playlist_updated", self._handle_playlist_updated)
            self._conn.on("player_online", lambda _: logger.debug("player_online"))
            self._conn.on("player_offline", lambda _: logger.debug("player_offline"))
            self._conn.on("command_ack", lambda _: logger.debug("command_ack"))
            self._conn.start()
            return True
        except Exception as e:
            logger.error("Failed to connect to CMS hub: %s", e)
            return False

    def _on_open(self) -> None:
        self._connected = True
        logger.info("CMS hub connected — subscribing as player")
        try:
            self._conn.send("SubscribePlayer", [self.screen_id, self.api_key])
        except Exception as e:
            logger.error("SubscribePlayer failed: %s", e)

    def _on_close(self) -> None:
        self._connected = False
        logger.warning("CMS hub connection closed")

    def _handle_command(self, data) -> None:
        """`data` is a list of args as delivered by signalrcore."""
        payload = data[0] if isinstance(data, list) and data else data
        if not isinstance(payload, dict):
            logger.warning("CMS command: unexpected payload type %s", type(payload))
            return
        logger.info("CMS command received: %s", payload.get("commandType"))
        if self._on_command:
            try:
                self._on_command(payload)
            except Exception as e:
                logger.exception("on_command handler raised: %s", e)

    def _handle_playlist_updated(self, data) -> None:
        payload = data[0] if isinstance(data, list) and data else data
        if not isinstance(payload, dict):
            payload = {}
        logger.info("CMS playlist_updated received")
        if self._on_playlist_updated:
            try:
                self._on_playlist_updated(payload)
            except Exception as e:
                logger.exception("on_playlist_updated handler raised: %s", e)

    def ack_command(self, command_id: str, success: bool, error: Optional[str] = None) -> None:
        if not self._connected or self._conn is None:
            logger.warning("ack_command skipped — not connected")
            return
        try:
            req = {"CommandId": command_id, "Success": success, "ErrorMessage": error}
            self._conn.send("AckCommand", [req])
        except Exception as e:
            logger.error("ack_command failed: %s", e)

    def stop(self) -> None:
        if self._conn is None:
            return
        try:
            self._conn.stop()
        except Exception:
            pass
        self._conn = None
        self._connected = False

File: player/chromeos/src/realtime/cmsControlClient.ts
/**
 * SignalR client for the CMS control hub at /hubs/cms.
 *
 * Connects with the screen API key as access_token; calls SubscribePlayer
 * after OnConnected (which BCrypt-verifies the key server-side). Forwards
 * `command` and `playlist_updated` events to the caller, and exposes
 * `ackCommand` for completion reporting.
 */
import * as signalR from '@microsoft/signalr';
import type { PlayerConfig } from '../config';

export interface RemoteCommandPayload {
  id: string;
  screenId: string;
  commandType: string;
  payloadJson: string | null;
  status: string;
  issuedAt: string | null;
}

export interface PlaylistUpdatedPayload {
  screenId?: string;
  playlistId?: string;
  version?: number;
  defaultChanged?: boolean;
}

export interface CmsControlCallbacks {
  onCommand: (cmd: RemoteCommandPayload) => void;
  onPlaylistUpdated: (evt: PlaylistUpdatedPayload) => void;
}

export class CmsControlClient {
  private connection: signalR.HubConnection | null = null;
  private readonly config: PlayerConfig;
  private readonly callbacks: CmsControlCallbacks;
  private stopped = false;

  constructor(config: PlayerConfig, callbacks: CmsControlCallbacks) {
    this.config = config;
    this.callbacks = callbacks;
  }

  async start(): Promise<void> {
    this.stopped = false;
    const baseUrl = this.config.serverUrl.replace(/\/+$/, '');

    this.connection = new signalR.HubConnectionBuilder()
      .withUrl(`${baseUrl}/hubs/cms`, {
        accessTokenFactory: () => this.config.apiKey,
      })
      .withAutomaticReconnect({
        nextRetryDelayInMilliseconds: (ctx) =>
          Math.min(1000 * Math.pow(2, ctx.previousRetryCount), 30_000),
      })
      .configureLogging(signalR.LogLevel.Warning)
      .build();

    this.connection.on('command', (cmd: RemoteCommandPayload) => {
      console.log(`[CmsHub] command: ${cmd.commandType}`);
      this.callbacks.onCommand(cmd);
    });

    this.connection.on('playlist_updated', (evt: PlaylistUpdatedPayload) => {
      console.log('[CmsHub] playlist_updated');
      this.callbacks.onPlaylistUpdated(evt);
    });

    this.connection.on('player_online', () => console.debug('[CmsHub] player_online'));
    this.connection.on('player_offline', () => console.debug('[CmsHub] player_offline'));
    this.connection.on('command_ack', () => console.debug('[CmsHub] command_ack'));

    this.connection.onreconnected(() => {
      console.log('[CmsHub] Reconnected — re-subscribing');
      this.subscribe().catch(console.error);
    });

    this.connection.onclose(() => {
      if (!this.stopped) {
        console.warn('[CmsHub] Closed — scheduling reconnect');
        setTimeout(() => this.start().catch(console.error), 5000);
      }
    });

    try {
      await this.connection.start();
      console.log('[CmsHub] Connected');
      await this.subscribe();
    } catch (err) {
      console.error('[CmsHub] Connect failed:', err);
      if (!this.stopped) {
        setTimeout(() => this.start().catch(console.error), 5000);
      }
    }
  }

  private async subscribe(): Promise<void> {
    if (this.connection?.state !== signalR.HubConnectionState.Connected) return;
    try {
      await this.connection.invoke('SubscribePlayer', this.config.screenId, this.config.apiKey);
      console.log('[CmsHub] SubscribePlayer OK');
    } catch (err) {
      console.error('[CmsHub] SubscribePlayer failed:', err);
    }
  }

  async ackCommand(commandId: string, success: boolean, errorMessage?: string): Promise<void> {
    if (this.connection?.state !== signalR.HubConnectionState.Connected) return;
    try {
      await this.connection.invoke('AckCommand', {
        CommandId: commandId,
        Success: success,
        ErrorMessage: errorMessage ?? null,
      });
    } catch (err) {
      console.warn('[CmsHub] AckCommand failed:', err);
    }
  }

  async stop(): Promise<void> {
    this.stopped = true;
    if (this.connection) {
      try {
        await this.connection.stop();
      } catch {
        // ignore
      }
      this.connection = null;
    }
  }
}
